### ISSUE TITLE
Fix `agent_upgrade` distributed behavior

### ISSUE THREAD

**Issue by s-ocando** [wazuh/wazuh#13846]
|Wazuh version|Install type|Install method|
|---|---|---|
| 4.3.4 |  Manager| Packages |

A worker is unable to see an outdated agent connected to the master when using `agent_upgrade -l`, it only lists the outdated agents connected to itself. 

When upgrading from the worker an agent that is connected to the master, the upgrade is successful but the script shows the following errors:
```
/var/ossec/bin/agent_upgrade -a 004
```

```
Upgrading...
Internal error: '004'
Traceback (most recent call last):
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 220, in <module>
    main()
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 197, in main
    check_status(affected_agents=result.affected_items, result_dict=agents_versions,
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 160, in check_status
    result_dict[task_result['agent']]['new_version'] = args.version if args.version else agent.version
KeyError: '004'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 227, in <module>
    if args.debug:
AttributeError: 'Namespace' object has no attribute 'debug'
```

**Comment by Pernat1y** [comment]
Same on OVA 4.3.5
```
# /var/ossec/bin/agent_upgrade --agents 045
Agent information not found in database
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/process.py", line 243, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.5-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 243, in run_local
    data = f(**f_kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.5-py3.9.egg/wazuh/rbac/decorators.py", line 420, in wrapper
    result = func(*args, **kwargs) if not skip_execution else None
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.5-py3.9.egg/wazuh/agent.py", line 981, in upgrade_agents
    raise WazuhInternalError(error_code, cmd_error=True, extra_message=agent_result['message'])
wazuh.core.exception.WazuhInternalError: Error 1816 - Agent information not found in database
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.5-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 288, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
wazuh.core.exception.WazuhInternalError: Error 1816 - Agent information not found in database
Internal error: 
Traceback (most recent call last):
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 220, in <module>
    main()
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 186, in main
    result = send_command(function=upgrade_agents, command=create_command())
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 115, in send_command
    return raise_if_exc(pool.submit(run, dapi.distribute_function()).result())
  File "/var/ossec/framework/python/lib/python3.9/site-packages/api-4.3.5-py3.9.egg/api/util.py", line 293, in raise_if_exc
    _create_problem(obj)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/api-4.3.5-py3.9.egg/api/util.py", line 263, in _create_problem
    raise ProblemException(status=500 if not code else code, type=exc.type, title=exc.title, detail=exc.message,
connexion.exceptions.ProblemException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 227, in <module>
    if args.debug:
AttributeError: 'Namespace' object has no attribute 'debug'
```

**Comment by davidjiglesias** [comment]
We should also check the behavior of agent_upgrade.py when the cluster is disabled.

**Comment by mcarmona99** [comment]
## Issue update

As @s-ocando said, when we try to upgrade an agent (or an agents list) using the `agent_upgrade` CLI from a worker node, an error is raised when that agent is connected to the master node (or to a different worker node).

Despite the error, the agent is upgraded correctly due to the fact that the exception is raised when trying to get the upgrade result.

The origin of the error is in the following function:

https://github.com/wazuh/wazuh/blob/cebdd5708b6759c84cbc409df7a670cf7f351619/framework/scripts/agent_upgrade.py#L52-L74

We cannot use this function because the worker's `global.db` does not contain information about agents connected to different nodes. The version must therefore be obtained with a distributed function.

Note: if the fix's target version is >=4.4, we must use the `forward_function` function (`core/cluster/utils.py`) to apply the fix. Note that in that case, we must also use the `forward_function` function instead of the `send_command` function in the rest of the CLI. 



Apart from fixing this error, we could also review the following code, as there is a possible race condition (very rare, but it could happen: the version is obtained after the upgrade procedure started):

https://github.com/wazuh/wazuh/blob/cebdd5708b6759c84cbc409df7a670cf7f351619/framework/scripts/agent_upgrade.py#L186-L194

Summarizing, this could be the ToDo list:

- Fix agent_upgrade -a XXX error
- Investigate and fix agent_upgrade -l error
- Use forward_function instead of send_command in agent_upgrade.py
- Fix rare race condition

## Checks

The following elements have been updated or reviewed (should also be checked if no modification is required):
- [ ] Tests (unit tests, API integration tests).
- [ ] Changelog.
- [ ] Documentation.
- [ ] Integration test mapping (using `api/test/integration/mapping/_test_mapping.py`).

**Comment by vicferpoy** [comment]
# Issue update

This bug has been fixed and properly tested. Users will be able to upgrade any agent from any cluster node and exceptions (if any) will be correctly handled.

```shellsession
root@wazuh-worker1:/# /var/ossec/bin/agent_upgrade -a 004 -v 4.3.10 -d

Upgrading...

Upgraded agents:
	Agent 004 upgraded: Wazuh v4.3.8 -> 4.3.10

root@wazuh-worker1:/# 
root@wazuh-worker1:/# 
root@wazuh-worker1:/# /var/ossec/bin/agent_upgrade -a 005 -v 4.4.0 -d

Upgrading...

Failed upgrades:
	Agent 005 status: The version of the WPK does not exist in the repository
root@wazuh-worker1:/# 
```
